File: web/static/ts3-connection.js
const ts3ConnFetch = typeof authFetch === 'function' ? authFetch : fetch;

function setTS3ConnStatus(message, isError = false) {
    const el = document.getElementById('ts3conn-status');
    if (!el) {
        return;
    }
    el.textContent = message;
    el.className = 'save-status ' + (isError ? 'err' : 'ok');
}

function formatTS3ConnectionError(status) {
    if (!status || !status.last_error) {
        return '-';
    }
    if (status.bot_running === false || status.last_error === 'bot disabled') {
        return t('ts3conn.botDisabledInfo');
    }
    return status.last_error;
}

function applyTS3ConnectionStatus(status) {
    const botRunning = status.bot_running !== false;
    const connected = !!status.connected;
    const connectedEl = document.getElementById('ts3conn-connected');
    if (connectedEl) {
        if (!botRunning) {
            connectedEl.textContent = t('ts3conn.botDisabled');
            connectedEl.style.color = 'var(--danger)';
        } else {
            connectedEl.textContent = connected ? t('common.yes') : t('common.no');
            connectedEl.style.color = connected ? 'var(--success)' : 'var(--danger)';
        }
    }

    const hostEl = document.getElementById('ts3conn-host');
    if (hostEl) {
        hostEl.textContent = status.host || '-';
    }

    const portEl = document.getElementById('ts3conn-port');
    if (portEl) {
        portEl.textContent = String(status.port || '-');
    }

    const checkEl = document.getElementById('ts3conn-check');
    if (checkEl) {
        checkEl.textContent = status.last_check_at || '-';
    }

    const errEl = document.getElementById('ts3conn-error');
    if (errEl) {
        errEl.textContent = formatTS3ConnectionError(status);
    }
}

async function loadTS3ConnectionStatus() {
    try {
        const res = await ts3ConnFetch('/api/ts3/connection/status');
        if (!res.ok) {
            const text = await res.text();
            throw new Error(text || 'status failed');
        }
        const payload = await res.json();
        applyTS3ConnectionStatus(payload.status || {});
        setTS3ConnStatus(t('ts3conn.statusLoaded'), false);
    } catch (err) {
        setTS3ConnStatus(t('ts3conn.statusLoadError', { error: err && err.message ? err.message : err }), true);
    }
}

async function runTS3ConnectionTest() {
    try {
        const res = await ts3ConnFetch('/api/ts3/connection/test', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' }
        });
        const payload = await res.json();
        applyTS3ConnectionStatus(payload.status || {});

        if (!payload.ok) {
			if (payload.status && payload.status.bot_running === false) {
				setTS3ConnStatus(t('ts3conn.botDisabledInfo'), true);
				return;
			}
            const err = payload.error || t('ts3conn.testFailed');
            setTS3ConnStatus(t('ts3conn.testError', { error: err }), true);
            return;
        }

        setTS3ConnStatus(t('ts3conn.testOk'), false);
    } catch (err) {
        setTS3ConnStatus(t('ts3conn.testError', { error: err && err.message ? err.message : err }), true);
    }
}

document.addEventListener('DOMContentLoaded', loadTS3ConnectionStatus);

File: web/api/ts3_connection.go
package api

import (
	"encoding/json"
	"net/http"
	"time"
)

type TS3ConnectionStatus struct {
	BotRunning     bool   `json:"bot_running"`
	Connected      bool   `json:"connected"`
	Host           string `json:"host"`
	Port           int    `json:"port"`
	LastCheckAt    string `json:"last_check_at"`
	LastError      string `json:"last_error,omitempty"`
	Implementation string `json:"implementation"`
}

var (
	GetTS3ConnectionStatusFunc func() TS3ConnectionStatus
	TestTS3ConnectionFunc      func() (TS3ConnectionStatus, error)
)

func TS3ConnectionStatusHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}

	if GetTS3ConnectionStatusFunc == nil {
		http.Error(w, "ts3 connection status unavailable", http.StatusServiceUnavailable)
		return
	}

	status := GetTS3ConnectionStatusFunc()
	if status.LastCheckAt == "" {
		status.LastCheckAt = time.Now().Format(time.RFC3339)
	}

	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(map[string]any{
		"ok":     true,
		"status": status,
	})
}

func TS3ConnectionTestHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}

	if TestTS3ConnectionFunc == nil {
		http.Error(w, "ts3 connection test unavailable", http.StatusServiceUnavailable)
		return
	}

	status, err := TestTS3ConnectionFunc()
	w.Header().Set("Content-Type", "application/json")
	if err != nil {
		_ = json.NewEncoder(w).Encode(map[string]any{
			"ok":     false,
			"error":  err.Error(),
			"status": status,
		})
		return
	}

	_ = json.NewEncoder(w).Encode(map[string]any{
		"ok":     true,
		"status": status,
	})
}
